### PR TITLE
Spherical Coordinate Averaging

### LINKED ISSUE #917: Better spectrum positions
For data converted from Mantid, scipp stores positions of spectra (not detectors). As a consequence the scattering angle computed from the position is not the average scattering angle of the underlying detectors. This is probably a minor error (to be verified).

The simplest (and best?) solution might be to compute a better spectrum position at the point of conversion from Mantid: The average position should preserve the average scattering angle, i.e., should probably be the average in a spherical coordinate system given by source and sample positions.

### PR DESCRIPTION
**Pull request by OwenArnold**
Fixes #917 

This needs to be merged first https://github.com/mantidproject/mantid/pull/28225

All detectors in spectrum are used to form spectrum average position in spherical coords.

### CODE REVIEW

File: python/src/scipp/compat/mantid.py [hunk 1]
```diff
@@ -195,15 +195,41 @@ def validate_and_get_unit(unit):
...
         if spec_info.hasDetectors(i):
-            p = spec_info.position(i)
-            pos[i, :] = [p.X(), p.Y(), p.Z()]
+            definition = spec_info.getSpectrumDefinition(i)
+            theta, phi, r = 0.0, 0.0, 0.0
+            n_dets = len(definition)
+            for j in range(n_dets):
+                p = det_info.position(definition[j][0])
+                x, y, z = np.array(p) - sample_pos.values
+                r_j, theta_j, phi_j = _to_spherical(x, y, z)
```
Comment (SimonHeybrock) [inline python/src/scipp/compat/mantid.py]: I suppose we should check that the primary flight path is aligned with Z, otherwise this does not make sense, or would at least give slight deviations, e.g., when source or sample are slightly offset.

File: python/src/scipp/compat/mantid.py [hunk 2]
```diff
@@ -195,15 +195,41 @@ def validate_and_get_unit(unit):
...
+    det_info = ws.detectorInfo()
+    pos = np.zeros([len(spec_info), 3])
+    for i in range(len(spec_info)):
         if spec_info.hasDetectors(i):
-            p = spec_info.position(i)
-            pos[i, :] = [p.X(), p.Y(), p.Z()]
+            definition = spec_info.getSpectrumDefinition(i)
+            theta, phi, r = 0.0, 0.0, 0.0
+            n_dets = len(definition)
+            for j in range(n_dets):
```
Comment (SimonHeybrock) [inline python/src/scipp/compat/mantid.py]: Isn't it really slow to do all the math in a loop? Can you just store the input values in an array?

Comment (OwenArnold) [reply python/src/scipp/compat/mantid.py]: `spec_info` is enumerable, so can make a small change around that, but cannot extract ALL positions from spectrum_info without loop. SpectrumDefinition likewise cannot get positions without going via DetectorInfo.

File: python/src/scipp/compat/mantid.py (hunk 1 above)
Comment (OwenArnold) [reply python/src/scipp/compat/mantid.py]: Yes, that could be done, could also change the definitions to fit around the beam axis.

File: python/src/scipp/compat/mantid.py (hunk 2 above)
Comment (SimonHeybrock) [reply python/src/scipp/compat/mantid.py]: Sure, you need the loop for the init, but all the more expensive math could be done on numpy arrays?

Comment (OwenArnold) [reply python/src/scipp/compat/mantid.py]: I pretty much had a working implementation for this, but it got really messy for the following reasons:

1. The spectra-detector mappings have to be dealt with. This means carrying around detector indexes and spectrum indexes and performing a grouping operation for the averaging at some point. 
1. Some spectra have no detectors. The processing route here is then quite different (we only know spectrum numbers, no grouping) so have to avoid grouping etc.
1. It is normally the case that the detector-spectra mapping is fixed, but this is not guaranteed. This affects the ability to preallocate arrays of the right side for the operations. At the very least a shrink is required. This is also true of the problems introduced by having no detectors.

Comment (SimonHeybrock) [reply python/src/scipp/compat/mantid.py]: Can you use `scipp.groupby` to do this in a nice way?

File: python/src/scipp/compat/mantid.py [hunk 3]
```diff
@@ -195,15 +195,70 @@ def validate_and_get_unit(unit):
...
-    for i in range(nHist):
+    det_info = ws.detectorInfo()
+    pos = np.zeros([len(spec_info), 3])
+    for i in range(len(spec_info)):
         if spec_info.hasDetectors(i):
-            p = spec_info.position(i)
-            pos[i, :] = [p.X(), p.Y(), p.Z()]
+            definition = spec_info.getSpectrumDefinition(i)
+            theta, phi, r = 0.0, 0.0, 0.0
+            n_dets = len(definition)
```
Comment (SimonHeybrock) [inline python/src/scipp/compat/mantid.py]: I am a bit worried about the performance implications. Can we just skip all the complicated bits in case there is just a single detector per spectrum? Have you checked the performance?

File: python/src/scipp/compat/mantid.py [hunk 4]
```diff
@@ -195,18 +195,126 @@ def validate_and_get_unit(unit):
...
+    pos_d["spectrum_idx"] = sc.Variable([Dim.X], values=spherical_[:, 0])
+    pos_d["r"] = sc.Variable([Dim.X], values=spherical_[:, 1], unit=sc.units.m)
+    pos_d["t"] = sc.Variable([Dim.X],
+                             values=spherical_[:, 2],
+                             unit=sc.units.rad)
+    pos_d["p"] = sc.Variable([Dim.X],
+                             values=spherical_[:, 3],
+                             unit=sc.units.rad)
+
+    grouping = sc.groupby(pos_d, "spectrum_idx", Dim.Y)
```
Comment (SimonHeybrock) [inline python/src/scipp/compat/mantid.py]: Remove `Dim.Y` for new syntax.

File: python/src/scipp/compat/mantid.py [hunk 5]
```diff
@@ -193,17 +193,119 @@ def validate_and_get_unit(unit):
...
+    s = np.linalg.norm(v)
+    if s == 0:
+        return np.eye(3)
+    kmat = np.array([[0, -v[2], v[1]], [v[2], 0, -v[0]], [-v[1], v[0], 0]])
+    rotation_matrix = np.eye(3) + kmat + kmat.dot(kmat) * ((1 - c) / (s**2))
+    return rotation_matrix
+
+
+def init_pos(ws, source_pos, sample_pos):
+    from scipp import Dim
```
Comment (SimonHeybrock) [inline python/src/scipp/compat/mantid.py]: Not necessary any more.

File: python/src/scipp/compat/mantid.py [hunk 6]
```diff
@@ -193,17 +193,119 @@ def validate_and_get_unit(unit):
...
     spec_info = ws.spectrumInfo()
-    for i in range(nHist):
-        if spec_info.hasDetectors(i):
-            p = spec_info.position(i)
-            pos[i, :] = [p.X(), p.Y(), p.Z()]
+    det_info = ws.detectorInfo()
+    total_detectors = spec_info.detectorCount()
+    non_detectors = sum(not spec.hasDetectors for spec in spec_info)
+    det_pos = np.zeros([total_detectors, 3 + 1])
+    no_detector_spectrum = np.full((non_detectors, 3 + 1), np.nan)
```
Comment (SimonHeybrock) [inline python/src/scipp/compat/mantid.py]: Why is this needed?

File: python/src/scipp/compat/mantid.py [hunk 7]
```diff
@@ -193,17 +193,119 @@ def validate_and_get_unit(unit):
...
+    positions[idx, :] = np.array([spec_idx, pos.X(), pos.Y(), pos.Z()])
+
+
+def _to_spherical(input):
+    transformed = np.ones(shape=input.shape)
+    transformed[:, 0] = input[:, 0]  # copy metadata
+    transformed[:, 1] = np.sqrt(np.sum(input[:, 1:]**2, axis=1))  # r
+    transformed[:, 2] = np.arccos(input[:, 3] /
+                                  transformed[:, 1])  # arccos ( z / r)
+    transformed[:, 3] = np.arctan2(input[:, 2], input[:, 1])  # arctan2(y, x)
```
Comment (SimonHeybrock) [inline python/src/scipp/compat/mantid.py]: Weird use of numpy arrays (why transposed?). Just use multiple outputs, or a dataset?

File: python/src/scipp/compat/mantid.py [hunk 8]
```diff
@@ -193,17 +193,119 @@ def validate_and_get_unit(unit):
...
+        values=np.array([np.nan] * len(no_detector_spectrum)),
+        unit=sc.units.m)
+    no_detector_ds["t"] = sc.Variable(
+        ["spectrum_idx"],
+        values=np.array([np.nan] * len(no_detector_spectrum)),
+        unit=sc.units.rad)
+    no_detector_ds["p"] = sc.Variable(
+        ["spectrum_idx"],
+        values=np.array([np.nan] * len(no_detector_spectrum)),
+        unit=sc.units.rad)
```
Comment (SimonHeybrock) [inline python/src/scipp/compat/mantid.py]: All the missing det handling seems overly complicated. Can't you just use a fake Id such as `-1` for missing dets, then `groupby`, and finally ignore the "-1" group?

File: python/src/scipp/compat/mantid.py (hunk 6 above)
Comment (OwenArnold) [reply python/src/scipp/compat/mantid.py]: There are some spectra that do not have any detectors. We ultimately still need to provide a position for those spectra.

File: python/src/scipp/compat/mantid.py (hunk 7 above)
Comment (OwenArnold) [reply python/src/scipp/compat/mantid.py]: Nothing is transposed here.

It's probably more convenient to do the transform as part of a scipp dataset.

File: python/src/scipp/compat/mantid.py (hunk 8 above)
Comment (OwenArnold) [reply python/src/scipp/compat/mantid.py]: Yes, that would slightly simplify the grouping. We still have to account for spectra with no detectors. We need to know what the indexes of those spectra are so that we can return a properly sorted result.

File: python/src/scipp/compat/mantid.py (hunk 6 above)
Comment (SimonHeybrock) [reply python/src/scipp/compat/mantid.py]: It should be `nan`, no? `groupby.mean` should just do this automatically. See last example in https://scipp.github.io/user-guide/groupby.html.

File: python/src/scipp/compat/mantid.py (hunk 7 above)
Comment (OwenArnold) [reply python/src/scipp/compat/mantid.py]: There are a few missing trig functions I need to add to `variable.cpp` to get this to work fully in scipp.

File: python/src/scipp/compat/mantid.py (hunk 8 above)
Comment (SimonHeybrock) [reply python/src/scipp/compat/mantid.py]: Why not something like:

1. Create table (dataset) with all detector information
2. loop over all spectra, write spectrum index for all detectors
3. `groupby(..., `spectrum-index`, bins=np.arange(-0.5, nspec + 0.5, 1.0))` (using `groupby` with `bins` to get NAN for spectra, instead of not having output.
4. Convert back to position and copy output arrays.

File: python/src/scipp/compat/mantid.py [hunk 9]
```diff
@@ -197,21 +197,120 @@ def validate_and_get_unit(unit, allow_empty=False):
...
+        act_beam = (sample_pos - source_pos).values
+        rot = rotation_matrix_from_vectors(act_beam, [0, 0, 1])
+        inv_rot = rot.transpose()
+    else:
+        rot = np.eye(3)
+        inv_rot = rot
+
+    pos_d = sc.Dataset()
+    # Create empty to hold position info for all spectra detectors
+    pos_d["x"] = sc.Variable(["x"],
```
Comment (SimonHeybrock) [inline python/src/scipp/compat/mantid.py]: Should the dimension be `"detector"`? Using `"x"` is a bit confusing (and makes the X-coord special, because it is now a dimension-coord (not sure this has any effect anywhere below).

File: python/src/scipp/compat/mantid.py [hunk 10]
```diff
@@ -197,21 +197,120 @@ def validate_and_get_unit(unit, allow_empty=False):
...
+    det_info = ws.detectorInfo()
+    total_detectors = spec_info.detectorCount()
+
+    if sample_pos and source_pos:
+        act_beam = (sample_pos - source_pos).values
+        rot = rotation_matrix_from_vectors(act_beam, [0, 0, 1])
+        inv_rot = rot.transpose()
+    else:
+        rot = np.eye(3)
+        inv_rot = rot
```
[the remark this answers is not in the document]
Comment (OwenArnold) [reply python/src/scipp/compat/mantid.py]: Yes I think that would be the right thing to do.

File: python/src/scipp/compat/mantid.py [hunk 11]
```diff
@@ -197,21 +197,127 @@ def validate_and_get_unit(unit, allow_empty=False):
...
+    if sample_pos and source_pos:
+        det_info = ws.detectorInfo()
+        total_detectors = spec_info.detectorCount()
+        act_beam = (sample_pos - source_pos).values
+        rot = rotation_matrix_from_vectors(act_beam, [0, 0, 1])
+        inv_rot = rot.transpose()
+
+        pos_d = sc.Dataset()
+        # Create empty to hold position info for all spectra detectors
+        pos_d["x"] = sc.Variable(["Detector"],
```
Comment (SimonHeybrock) [inline python/src/scipp/compat/mantid.py]: We have standardized to lower-case dimension names.
```suggestion
        pos_d["x"] = sc.Variable(["detector"],
```

File: python/src/scipp/compat/mantid.py [hunk 12]
```diff
@@ -197,21 +197,127 @@ def validate_and_get_unit(unit, allow_empty=False):
...
+
+        _to_spherical(pos_d)
+
+        averaged = sc.groupby(pos_d,
+                              "spectrum",
+                              bins=sc.Variable(["spectrum"],
+                                               values=np.arange(
+                                                   -0.5,
+                                                   len(spec_info) + 0.5,
+                                                   1.0))).mean("Detector")
```
Comment (SimonHeybrock) [inline python/src/scipp/compat/mantid.py]:
```suggestion
                                                   1.0))).mean("detector")
```

File: python/src/scipp/compat/mantid.py [hunk 13]
```diff
@@ -197,21 +197,127 @@ def validate_and_get_unit(unit, allow_empty=False):
...
 
 
-def init_pos(ws):
-    nHist = ws.getNumberHistograms()
-    pos = np.zeros([nHist, 3])
-
+def _to_spherical(input):
+    input["r"] = sc.sqrt(input["x"].data * input["x"].data +
+                         input["y"].data * input["y"].data +
+                         input["z"].data * input["z"].data)
```
Comment (SimonHeybrock) [inline python/src/scipp/compat/mantid.py]: Any reason for not using https://scipp.github.io/generated/scipp.dot.html#scipp.dot based on the position?

File: python/src/scipp/compat/mantid.py [hunk 14]
```diff
@@ -197,21 +197,127 @@ def validate_and_get_unit(unit, allow_empty=False):
...
+    kmat = np.array([[0, -v[2], v[1]], [v[2], 0, -v[0]], [-v[1], v[0], 0]])
+    rotation_matrix = np.eye(3) + kmat + kmat.dot(kmat) * ((1 - c) / (s**2))
+    return rotation_matrix
+
+
+def matrix_mult(pos, m):
+    if not pos.shape[0]:
+        return pos
+    m1 = sc.Variable(['y'],
+                     values=[np.array(m[0, :])],
```
Comment (SimonHeybrock) [inline python/src/scipp/compat/mantid.py]: Why are the extra `[]` required around `np.array`?

Comment (OwenArnold) [reply python/src/scipp/compat/mantid.py]: I think this is a limitation of the Variable constructor. Cannot initialise with a single value for vector_3_float64. I mean to open an issue up about that.

File: python/src/scipp/compat/mantid.py (hunk 13 above)
Comment (OwenArnold) [reply python/src/scipp/compat/mantid.py]: Good point.

File: python/src/scipp/compat/mantid.py [hunk 15]
```diff
@@ -197,21 +197,127 @@ def validate_and_get_unit(unit, allow_empty=False):
...
+                     dtype=sc.dtype.vector_3_float64)
+    m2 = sc.Variable(['y'],
+                     values=[np.array(m[1, :])],
+                     dtype=sc.dtype.vector_3_float64)
+    m3 = sc.Variable(['y'],
+                     values=[np.array(m[2, :])],
+                     dtype=sc.dtype.vector_3_float64)
+    xn = sc.sum(sc.dot(m1, pos), 'y')
+    yn = sc.sum(sc.dot(m2, pos), 'y')
+    zn = sc.sum(sc.dot(m3, pos), 'y')
```
Comment (SimonHeybrock) [inline python/src/scipp/compat/mantid.py]: What is the significance of `y` here and in the definitions of `m1`, ...? 
`pos` depends on `'detector'`, so the results of the `dot` are 2D arrays?

File: python/src/scipp/compat/mantid.py (hunk 14 above)
Comment (SimonHeybrock) [reply python/src/scipp/compat/mantid.py]: Have you tried `value=` instead of `values=`?

Comment (SimonHeybrock) [reply python/src/scipp/compat/mantid.py]: I think this also explains my question below regarding `'y'`. Remove it here and below, and use `value=`?

File: python/src/scipp/compat/mantid.py [hunk 16]
```diff
@@ -197,21 +197,127 @@ def validate_and_get_unit(unit, allow_empty=False):
...
-            p = spec_info.position(i)
-            pos[i, :] = [p.X(), p.Y(), p.Z()]
-        else:
-            pos[i, :] = [np.nan, np.nan, np.nan]
-    return sc.Variable(['spectrum'],
-                       values=pos,
-                       unit=sc.units.m,
-                       dtype=sc.dtype.vector_3_float64)
+
+    if sample_pos and source_pos:
```
Comment (SimonHeybrock) [inline python/src/scipp/compat/mantid.py]: Is this the correct way of checking in Python? Aren't these `None`?

File: python/src/scipp/compat/mantid.py (hunk 14 above)
Comment (OwenArnold) [reply python/src/scipp/compat/mantid.py]: I stand corrected. ```sc.Variable(value=np.array([1,2,3]),dtype=sc.dtype.vector_3_float64)``` does actually work.

File: python/tests/compat/test_mantid.py [hunk 17]
```diff
@@ -100,8 +100,10 @@ def test_unit_conversion(self):
         converted_mantid = mantidcompat.convert_Workspace2D_to_data_array(ws)
 
         da = mantidcompat.convert_EventWorkspace_to_data_array(eventWS, False)
-        da = sc.histogram(da, target_tof)
-        d = sc.Dataset(da)
+        da_histo = sc.histogram(da, target_tof)
+        da_histo.coords['source_position'] = da.coords['source_position']
+        da_histo.coords['sample_position'] = da.coords['sample_position']
+        d = sc.Dataset(da_histo)
```
Comment (SimonHeybrock) [inline python/tests/compat/test_mantid.py]: Why this change? It looks like a workaround for a bug that is already fixed?
